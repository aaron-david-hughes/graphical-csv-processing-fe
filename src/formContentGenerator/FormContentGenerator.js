import React from 'react';
import Input from "../generalPurposeComponents/input/Input";
import Switch from "../generalPurposeComponents/switch/Switch";

class FormContentGenerator extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            nodeOperation: props.nodeOperation,
            inputActionFunctions: props.inputActionFunctions,
            isSettings: !!props.isSettings
        }
    }

    static getDerivedStateFromProps(props, state) {
        if (props.nodeOperation !== state.nodeOperation) {
            return {
                ...state,
                nodeOperation: props.nodeOperation
            }
        }

        return null;
    }

    generateField(field, fieldConfig) {
        switch (fieldConfig.input) {
            case 'text':
                return this.textField(field, fieldConfig);
            case 'numeric':
                return this.numericField(field, fieldConfig);
            case 'integer':
                return this.integerField(field, fieldConfig);
            case 'dropdown':
                return this.dropdownField(field, fieldConfig);
            case 'switch':
                return this.switchField(field, fieldConfig);
            default:
                console.log('input unknown');
        }
    }

    async onBlur(onBlur, e, field, fieldConfig) {
        return this.state.isSettings
            ? onBlur(e, this.props.nodeOperation, field)
            : onBlur(e, field, fieldConfig.required)
    }

    async onChange(onChange, e, field) {
        this.state.isSettings
            ? onChange(e, this.props.nodeOperation, field)
            : onChange(e, field)
    }

    textField(field, fieldConfig) {
        let id = this.props.nodeOperation + '-' + field;
        let defaultValue = null;

        if (this.props.node) {
            defaultValue = this.props.node[field]
        }

        if (this.props.nodeObj) {
            defaultValue = this.props.nodeObj[field]
        }

        return <Input
            key={id}
            id={id}
            errorText={fieldConfig.errorText ? fieldConfig.errorText : 'Please fill in this field.'}
            title={fieldConfig.title}
            isInvalid={this.props.inputValidity[field] === 'invalid'}
            style={{width: fieldConfig.width ? fieldConfig.width : '100%'}}
            required={fieldConfig.required}
            input={
                <input
                    id={id + '-input'}
                    type='text'
                    defaultValue={defaultValue}
                    onBlur={async e => this.onBlur(this.state.inputActionFunctions['text'].onBlur, e, field, fieldConfig)}
                />
            }
        />
    }

    numericField(field, fieldConfig) {
        let id = this.props.nodeOperation + '-' + field;
        let defaultValue = null;

        if (this.props.node) {
            defaultValue = this.props.node[field]
        }

        if (this.props.nodeObj) {
            defaultValue = this.props.nodeObj[field]
        }

        return <Input
            key={id}
            id={id}
            errorText={fieldConfig.errorText ? fieldConfig.errorText : 'Please supply a number.'}
            title={fieldConfig.title}
            isInvalid={this.props.inputValidity[field] === 'invalid'}
            style={{width: fieldConfig.width ? fieldConfig.width : '100%'}}
            required={fieldConfig.required}
            input={
                <input
                    id={id + '-input'}
                    type='text'
                    defaultValue={defaultValue}
                    onBlur={async e => this.onBlur(this.state.inputActionFunctions['numeric'].onBlur, e, field, fieldConfig)}
                />
            }
        />
    }

    integerField(field, fieldConfig) {
        let id = this.props.nodeOperation + '-' + field;
        let defaultValue = null;

        if (this.props.node) {
            defaultValue = this.props.node[field]
        }

        if (this.props.nodeObj) {
            defaultValue = this.props.nodeObj[field]
        }

        return <Input
            key={id}
            id={id}
            errorText={fieldConfig.errorText ? fieldConfig.errorText : 'Please supply a integer.'}
            title={fieldConfig.title}
            isInvalid={this.props.inputValidity[field] === 'invalid'}
            style={{width: fieldConfig.width ? fieldConfig.width : '100%'}}
            required={fieldConfig.required}
            input={
                <input
                    id={id + '-input'}
                    type='text'
                    defaultValue={defaultValue}
                    onBlur={async e => this.onBlur(this.state.inputActionFunctions['integer'].onBlur, e, field, fieldConfig)}
                />
            }
        />
    }

    dropdownField(field, fieldConfig) {
        let id = this.props.nodeOperation + '-' + field;
        let defaultValue = fieldConfig.options[0];

        if (this.props.node) {
            defaultValue = this.props.node[field]
        }

        if (this.props.nodeObj) {
            defaultValue = this.props.nodeObj[field]
        }

        return <Input
            key={id}
            id={id}
            title={fieldConfig.title}
            style={{width: fieldConfig.width ? fieldConfig.width : '100%'}}
            required={fieldConfig.required}
            input={
                <select
                    id={id + '-input'}
                    style={{
                        width: '100%',
                        padding: '1px 2px'
                    }}
                    defaultValue={defaultValue}
                    onChange={async e => this.onChange(this.state.inputActionFunctions['dropdown'].onChange, e, field)}
                >
                    {
                        fieldConfig.options.map(function(type) {
                            return <option key={type} value={type}>
                                {type}
                            </option>
                        })
                    }
                </select>
            }
        />
    }

    switchField(field, fieldConfig) {
        let id = this.props.nodeOperation + '-' + field;
        return <Input
            key={id}
            id={id}
            title={fieldConfig.title}
            style={{width: fieldConfig.width ? fieldConfig.width : '100%'}}
            required={fieldConfig.required}
            input={
                <Switch
                    id={id + '-input'}
                    isChecked={this.props.nodeObj[field]}
                    onChange={async e => this.onChange(this.state.inputActionFunctions['switch'].onChange, e, field)}
                />
            }
        />
    }

    processElementContent() {
        let operationConfig = this.props.config.processing.operations.find(op => op.operation === this.state.nodeOperation);

        return <div
            id={'processingForm-' + this.state.nodeOperation}
            style={{
                display: 'flex',
                flexWrap: 'wrap',
                width: '100%'
            }}
        >
            {
                Object.keys(operationConfig.template)
                    .filter(field => field !== 'operation' && field !== 'expectedInputs')
                    .filter(field => operationConfig[field] !== undefined)
                    .map(field => this.generateField(field, operationConfig[field]))
            }
        </div>
    }

    render() {
        return this.processElementContent();
    }
}

export default FormContentGenerator;