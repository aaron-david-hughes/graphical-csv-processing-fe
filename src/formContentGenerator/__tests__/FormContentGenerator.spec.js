import React from 'react';
import {shallow, mount} from 'enzyme';
import FormContentGenerator from "../FormContentGenerator";

describe('FormContentGenerator tests', () => {
    let props = {
        nodeOperation: 'testOp',
        inputActionFunctions: {
            text: {
                onBlur: jest.fn()
            },
            numeric: {
                onBlur: jest.fn()
            },
            integer: {
                onBlur: jest.fn()
            },
            dropdown: {
                onChange: jest.fn()
            },
            switch: {
                onChange: jest.fn()
            }
        },
        inputValidity: {
            textFieldName: 'valid',
            numericFieldName: 'valid',
            integerFieldName: 'valid'
        },
        nodeObj: {
            switchFieldName: true
        },
        config: {
            processing: {
                operations: [{
                    operation: 'testOp',
                    name: 'Test Operation',
                    template: {
                        operation: 'testOp',
                        textFieldName: null,
                        numericFieldName: null,
                        integerFieldName: null,
                        dropdownFieldName: 'first',
                        switchFieldName: true,
                        expectedInputs: 2
                    },
                    textFieldName: {
                        input: 'text',
                        title: 'Text Field Name',
                        width: "100%",
                        required: true
                    },
                    numericFieldName: {
                        input: 'numeric',
                        title: 'Numeric Field Name',
                        width: "100%",
                        required: true
                    },
                    integerFieldName: {
                        input: 'integer',
                        title: 'Integer Field Name',
                        width: "100%",
                        required: true
                    },
                    dropdownFieldName: {
                        input : 'dropdown',
                        title: 'Dropdown Field Name',
                        width: '100%',
                        required: true,
                        options: [
                            'first',
                            'second',
                            'third'
                        ]
                    },
                    switchFieldName: {
                        input: 'switch',
                        title: 'Switch Field Name',
                        width: '100%',
                        required: true
                    }
                }]
            }
        }
    };

    let render;

    beforeEach(() => {
        render = shallow(<FormContentGenerator {...props} />);
    });

    afterEach(() => {
        render.unmount();
    });

    describe('FormContentGenerator render tests', () => {
        it('should match snapshot', () => {
            expect(render).toMatchSnapshot();
        });

        it('should call text onBlur function when text field blur event occurs', () => {
            render = mount(<FormContentGenerator {...props} />);
            render.find('#testOp-textFieldName-input').simulate('blur');

            expect(props.inputActionFunctions.text.onBlur).toHaveBeenCalled();
        });

        it('should call numeric onBlur function when numeric field blur event occurs', () => {
            render = mount(<FormContentGenerator {...props} />);
            render.find('#testOp-numericFieldName-input').simulate('blur');

            expect(props.inputActionFunctions.numeric.onBlur).toHaveBeenCalled();
        });

        it('should call integer onBlur function when integer field blur event occurs', () => {
            render = mount(<FormContentGenerator {...props} />);
            render.find('#testOp-integerFieldName-input').simulate('blur');

            expect(props.inputActionFunctions.integer.onBlur).toHaveBeenCalled();
        });

        it('should call dropdown onChange function when dropdown change event occurs', () => {
            render = mount(<FormContentGenerator {...props} />);
            render.find('#testOp-dropdownFieldName-input').simulate('change');

            expect(props.inputActionFunctions.dropdown.onChange).toHaveBeenCalled();
        });
    });

    describe('FormContentGenerator instance tests', () => {
        it('should update state when nodeOperation prop supplied changes', () => {
            let instance = render.instance();

            let newProps = {
                nodeOperation: 'newTestOp'
            }

            expect(instance.state.nodeOperation).toEqual('testOp');

            let newState = FormContentGenerator.getDerivedStateFromProps(newProps, render.state);

            expect(newState.nodeOperation).toEqual('newTestOp');
        });
    });
});