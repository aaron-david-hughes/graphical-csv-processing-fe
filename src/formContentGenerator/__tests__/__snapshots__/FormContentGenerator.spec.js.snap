// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormContentGenerator tests FormContentGenerator render tests should match snapshot 1`] = `
<div
  id="processingForm-testOp"
  style={
    Object {
      "display": "flex",
      "flexWrap": "wrap",
      "width": "100%",
    }
  }
>
  <Input
    errorText="Please fill in this field."
    id="testOp-textFieldName"
    input={
      <input
        id="testOp-textFieldName-input"
        onBlur={[Function]}
        type="text"
      />
    }
    isInvalid={false}
    key="testOp-textFieldName"
    required={true}
    style={
      Object {
        "width": "100%",
      }
    }
    title="Text Field Name"
  />
  <Input
    errorText="Please supply a number."
    id="testOp-numericFieldName"
    input={
      <input
        id="testOp-numericFieldName-input"
        onBlur={[Function]}
        type="text"
      />
    }
    isInvalid={false}
    key="testOp-numericFieldName"
    required={true}
    style={
      Object {
        "width": "100%",
      }
    }
    title="Numeric Field Name"
  />
  <Input
    errorText="Please supply a integer."
    id="testOp-integerFieldName"
    input={
      <input
        id="testOp-integerFieldName-input"
        onBlur={[Function]}
        type="text"
      />
    }
    isInvalid={false}
    key="testOp-integerFieldName"
    required={true}
    style={
      Object {
        "width": "100%",
      }
    }
    title="Integer Field Name"
  />
  <Input
    id="testOp-dropdownFieldName"
    input={
      <select
        id="testOp-dropdownFieldName-input"
        onChange={[Function]}
        style={
          Object {
            "padding": "1px 2px",
            "width": "100%",
          }
        }
      >
        <option
          value="first"
        >
          first
        </option>
        <option
          value="second"
        >
          second
        </option>
        <option
          value="third"
        >
          third
        </option>
      </select>
    }
    key="testOp-dropdownFieldName"
    required={true}
    style={
      Object {
        "width": "100%",
      }
    }
    title="Dropdown Field Name"
  />
  <Input
    id="testOp-switchFieldName"
    input={
      <Switch
        id="testOp-switchFieldName-input"
        isChecked={true}
        onChange={[Function]}
      />
    }
    key="testOp-switchFieldName"
    required={true}
    style={
      Object {
        "width": "100%",
      }
    }
    title="Switch Field Name"
  />
</div>
`;
